MNT: add linter for thread-unsafe C API uses

lvllvl opened the pull request:
Adding a linter to audit PRs. 
This would lint out PRs that use problematic functions such as
- `PyList_GetItem`
- `PyList_GET_ITEM`
- `PyDict_GetItem`
- `PyDict_getItemWithError`
- `PyDict_Next`
- `PyDict_GetItemString`
- `_PyDict_GetItemStringWithError`

Attempts to resolve #26159 

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

ngoldbaum commented on .github/workflows/c_linter.yml:
```diff
@@ -0,0 +1,16 @@
+  pull_request:
+  workflow_dispatch:
+
+jobs:
```
Most of the other workflows have some code here to make sure the workflow doesn't run on forks unless someone explicitly opts into that by commenting out a line.

ngoldbaum commented:
Does this need a whole new CI job? I know I suggested that in the issue, but it occurs to me that this could be integrated into e.g. the `spin lint` command.

Also rather than just checking diffs we could do this codebase-wide and add e.g. `//  noqa` comments throughout the codebase, since it is safe to use borrowed references in some situation, but I think having a linter that alerts contributors and reviewers about the problem is good.

If the lint fails, it should suggest what to do to fix the issue. One of:

* use an API that returns a strong reference

or, add an opt-out to the lint if:

* you are very sure that a lock is held 
* you are very sure you have the only reference to the object (e.g. it is newly created in the local scope).

The linter should also probably check all of the functions listed in the free-threaded HOWTO in the CPython docs:

https://docs.python.org/3/howto/free-threading-extensions.html#borrowed-references

Also see here for more details about why we need to do this and why we can't just globally replace all these functions:

https://py-free-threading.github.io/porting-extensions/#cpython-c-api-usage

We probably should also be linting uses of fast accessor macros that don't do any locking.

Also, FWIW, there are probably still lurking issues. I'm aware of `Py_SEQUENCE_FAST` at least (https://github.com/numpy/numpy/issues/28046), I've just had more pressing things to work on and haven't been able to finish that one off. I'm hoping to finish off that work soon, or at least upstream the trivial fixes I have but leave an issue open for the more complicated issues in the array coercion code.

ngoldbaum commented on tools/ci/check_c_api_usage.sh:
```diff
@@ -0,0 +1,47 @@
+        # Run grep, if found, record and flag
+        if grep -n "$func" "$file" > /dev/null 2>&1; then
+
+            # If not whitelisted...
```
not seeing any way to whitelist functions, am I missing something?

lvllvl replied:
originally, I was assuming there'd be a whitelist of functions but I removed this idea from the script.

ngoldbaum commented on tools/ci/check_c_api_usage.sh:
```diff
@@ -0,0 +1,47 @@
+        if grep -n "$func" "$file" > /dev/null 2>&1; then
+
+            # If not whitelisted...
+            echo "Found suspicious call $func in $file" >> $OUTPUT
```
this error message should have more detail about what needs to happen to fix the issue

mhvk reviewed (commented):
I think this goes in the right direction, but may need a bit of care. Overall, perhaps good to keep it simple, and maybe just use `git grep` for all, like,
```
git grep -e PyList_GetItem\( --or -e PyList_GET_ITEM\( ... origin/main...HEAD
```
Or perhaps better write the list to a file and use `-f`.

Notes:
1. important to add the `(` otherwise correct functions like `PyList_GetItemRef` will be matched too.
2. You have to be sure to match *all* C files, so also `.c.src` and `.cpp`, etc. My suggestion would be to just match all files and *perhaps* remove `.rst`... since this only runs on the differences, false positives are not very likely.

seberg commented:
I almost wonder if one could do a custom clang-tidy style check (not if that can that work with `.c.src` though, I admit)?  That might also be useful for other projects then.

Plus that means we have an established pattern for allowing it, just add a `// NOLINT: Newly created tuple` comment.  I think an inline code-comment is vastly preferential to some global allowlist.

ngoldbaum commented:
This is looking much closer to being mergeable. Let me know if you need a hand with getting the circleci build fixed.

ngoldbaum commented:
.. or maybe it's unrelated and would go away with a rebase?

lvllvl commented:
> 

1. I think the rebase worked.
2. The test is set to fail if any of the functions are used but currently I think there's about 92 uses of the functions, [here's a link: C API Borrowed Ref Lint](https://github.com/numpy/numpy/actions/runs/14422363140/job/40446527735?pr=28634#step:3:120)
3. Should I assume that all the currently flagged uses of the functions are approved uses? I.e., should I just mark each listed instance as ok `// noqa: borrowed-ref OK`
4. I think a failing version of this test would cause 4 tests to fail, `azure-pipeline numpy`, `azure-pipeline numpy (ComprehensiveTest Lint)`, `Linux`, and `Run MyPy/C API Borrowed Ref Lint`. I think it's because those jobs `run linter.py`, is that ok?

ngoldbaum commented on numpy/_core/src/common/pythoncapi_compat.h:
This file should not have been added to the NumPy repo. Can you please squash the commits that add and remove it too?

ngoldbaum replied:
> Sorry I was just trying to get the tests to all pass so I added this from the submodule.

No worries, sorry if my tone was harsh, that wasn't what I was going for. As a sanity check when I review PRs I look at the total lines added and deleted on a PR and if it's really out of whack I start to look for stuff like this.

lvllvl replied:
No offense taken at all, thanks though. And thanks for the review. I should be able to address these changes this afternoon.

lvllvl commented:
@ngoldbaum looks like all the tests are passing now. 

1. Re: `numpy/_core/src/common/pythoncapi_compat.h`, I think I accidentally added this file into this location. So you were right, this was the source of all the additional lines. I removed this from my PR
2. Re: Linter in `tools/ci/check_c_api_usage.sh`, I changed the ALL_FILES filter to ignore the submodule section. I tested it in my codespace and here, seems to work ok. The line is now: 
```bash
ALL_FILES=$(find numpy -type f \( -name "*.c" -o -name "*.h" -o -name "*.c.src" -o -name "*.cpp" \) ! -path "*/pythoncapi-compat/*")
```

Let me know if you need any additional changes.

ngoldbaum commented on numpy/_core/src/common/ufunc_override.c:
```diff
@@ -108,7 +108,7 @@ PyUFuncOverride_GetOutObjects(PyObject *kwds, PyObject **out_kwd_obj, PyObject *
          */
         PyObject *seq;
-        seq = PySequence_Fast(*out_kwd_obj,
+        seq = PySequence_Fast(*out_kwd_obj, // noqa: borrowed-ref OK
```
I actually have some changes I need to upstream related to this from back in February, see https://github.com/numpy/numpy/issues/28046#issuecomment-2683109891. Maybe I should finally get around to sending in what I have...

ngoldbaum reviewed (commented):
Awesome, I'm glad this is shaping up.

I guess my main issue with this is that we should be relatively sure that all of the ones we're marking as OK really are OK, otherwise we'll be misleading future readers.

I need to sit down with this PR and try to stare at all the uses and see if I can come up with possible problems.

We might need a new NOQA category for uses that are known to be problematic but need manual fixes. That would allow us to catch and triage new uses in CI while not needing to go through and fix absolutely everything that's still in the library.

I will try to do this soon but may not have a ton of time before PyCon.

ngoldbaum commented:
Hi all, I'm really sorry for the delay here. Totally my fault for letting it sit for so long.

I'll try to set aside some time to get this up to shape.

ngoldbaum commented:
> How are you determining which use of free threading is acceptable and which ones are not? Is there criteria beyond the literature that you linked in earlier comments? I'll try and verify each is labeled as well.

The two links I shared above are where you should look. You could also watch the PyCon talk @lysnikolaou and I gave about free-threaded support: https://www.youtube.com/watch?v=EuU3ksI1l04, which covers this.

In short, borrowed references to items in mutable containers (e.g. dicts, lists) are unsafe if the container is visible to another thread, because the item might be de-allocated if its refcount happens to go to zero. Whether or not the container is visible to other threads is context-dependent. 

We don't want to wholesale move everything to strong reference APIs because of performance and the risk of bugs. Performance because borrowed reference skip an incref and a decref, which might be very expensive in a tight loop. Bugs because reference counting logic in C is very tricky, particularly in code that has complicated control flow, and is easy to screw up.

> I can add in an additional category to the script, e.g., label = noqa: manual fix needed and allow those noqa labeled comments to pass this linter as well.

Makes sense to me.

lvllvl committed:
Merge branch 'numpy:main' into issue-26159

ngoldbaum commented on numpy/_core/src/multiarray/methods.c:
```diff
@@ -1005,7 +1005,7 @@ any_array_ufunc_overrides(PyObject *args, PyObject *kwds)
         return -1;
     }
-    fast = PySequence_Fast(args, "Could not convert object to sequence");
+    fast = PySequence_Fast(args, "Could not convert object to sequence"); // noqa: borrowed-ref OK
```
I *think* this might be effectively private to NumPy?

lvllvl replied:
Ok sounds good. I'll leave this comment as is, then.

ngoldbaum commented on numpy/_core/src/multiarray/methods.c:
```diff
@@ -1733,7 +1733,7 @@ _setlist_pkl(PyArrayObject *self, PyObject *list)
     }
     while(iter->index < iter->size) {
-        theobject = PyList_GET_ITEM(list, iter->index);
+        theobject = PyList_GET_ITEM(list, iter->index); // noqa: borrowed-ref - manual fix needed
```
this list is effectively private to the calling function, so this is fine

ngoldbaum reviewed (commented):
spotted some issues with how you've set the markers so far, let me know if you have questions about my logic

lvllvl commented:
> spotted some issues with how you've set the markers so far, let me know if you have questions about my logic

cool thank you for reviewing. I had a loose rubric for each use of thread-unsafe functions but your input was helpful. 

1. Rubric was: Is container inmutable? Is the container new/private? Is there an explicit lock on container? Is there a guarantee that container will not be deleted? And if I answered "no" on all those questions, then I changed comment to `noqa: borrowed-ref - manual fix needed`
2. I also added all the labels from the [Borrowed reference API](https://docs.python.org/3/howto/free-threading-extensions.html#borrowed-references) so the linter will catch all those listed here
3. I'm not sure why I'm failing tests, sorry will investigate this asap

ngoldbaum commented on numpy/_core/src/multiarray/buffer.c:
```diff
@@ -268,7 +268,7 @@ _buffer_format_string(PyArray_Descr *descr, _tmp_string_t *str,
 
             name = PyTuple_GET_ITEM(ldescr->names, k);
-            item = PyDict_GetItem(ldescr->fields, name);
+            item = PyDict_GetItem(ldescr->fields, name); // noqa: borrowed-ref - manual fix needed
```
It looks like you missed an access to `fields` here. Can you go through and make sure all of them are marked OK per my explanation from the last round of review?

ngoldbaum commented on numpy/_core/src/multiarray/compiled_base.c:
```diff
@@ -1522,7 +1522,7 @@ arr_add_docstring(PyObject *NPY_UNUSED(dummy), PyObject *const *args, Py_ssize_t
         _ADDDOC(new->tp_doc, new->tp_name);
         if (new->tp_dict != NULL && PyDict_CheckExact(new->tp_dict) &&
-                PyDict_GetItemString(new->tp_dict, "__doc__") == Py_None) {
+                PyDict_GetItemString(new->tp_dict, "__doc__") == Py_None) { // noqa: borrowed-ref - manual fix needed
```
the instance passed into this function should have a strong reference to the type object, which has a strong reference to `tp_dict`, so this is OK

ngoldbaum commented on numpy/_core/src/multiarray/conversion_utils.c:
```diff
@@ -130,7 +130,7 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
      */
     if (!PyLong_CheckExact(obj) && PySequence_Check(obj)) {
-        seq_obj = PySequence_Fast(obj,
+        seq_obj = PySequence_Fast(obj, // noqa: borrowed-ref OK
```
I think this isn't OK, because another thread can access `obj`.

ngoldbaum commented on numpy/_core/src/multiarray/conversion_utils.c:
```diff
@@ -1135,7 +1135,7 @@ PyArray_IntpFromSequence(PyObject *seq, npy_intp *vals, int maxvals)
     PyObject *seq_obj = NULL;
     if (!PyLong_CheckExact(seq) && PySequence_Check(seq)) {
-        seq_obj = PySequence_Fast(seq,
+        seq_obj = PySequence_Fast(seq, // noqa: borrowed-ref OK
```
ditto

ngoldbaum commented on numpy/_core/src/multiarray/convert_datatype.c:
```diff
@@ -2749,7 +2749,7 @@ nonstructured_to_structured_resolve_descriptors(
             Py_ssize_t pos = 0;
             PyObject *key, *tuple;
-            while (PyDict_Next(to_descr->fields, &pos, &key, &tuple)) {
+            while (PyDict_Next(to_descr->fields, &pos, &key, &tuple)) { // noqa: borrowed-ref - manual fix needed
```
fields again, this is fine, ditto for all the other ones that go through PyDataType_FIELDS below.

ngoldbaum reviewed (commented):
I left a few comments below. Sorry I didn't make it clearer on the last review that I wasn't exhaustive in my comments. I'm also not exhaustive in pointing out every instance of a pattern I commented about here, please make sure that *all* of the ones dealing with `fields` below are marked as OK, since it's effectively append-only and I don't think it's worth the reference counting churn to fix every spot you have currently marked as needing a fix.

lvllvl commented on numpy/_core/src/multiarray/compiled_base.c:
```diff
@@ -1522,7 +1522,7 @@ arr_add_docstring(PyObject *NPY_UNUSED(dummy), PyObject *const *args, Py_ssize_t
         _ADDDOC(new->tp_doc, new->tp_name);
         if (new->tp_dict != NULL && PyDict_CheckExact(new->tp_dict) &&
-                PyDict_GetItemString(new->tp_dict, "__doc__") == Py_None) {
+                PyDict_GetItemString(new->tp_dict, "__doc__") == Py_None) { // noqa: borrowed-ref - manual fix needed
```
ok sounds good. Not sure what happened to my branch. Triaging atm, will try to re-open. Not sure why the PR closed when I committed.

lvllvl replied:
Hi @ngoldbaum, really sorry for the inconvenience but are you able to re-open this PR on your end? I am unable to re-open the PR.

I'm not sure why my `git push --force` automatically closed the PR. But I must've done something incorrectly. I'm reviewing my branch again to implement your suggested changes.

If re-opening on your end is not an option, I can submit a new PR. Again, sorry for the inconvenience.
